dryrun_diff: add HTML breakdown

JuubiSnake opened the pull request:
## Description

This PR adds a HTML rendered document that details a breakdown of affected policies between a dryrun and the last live run of manhiem. 

The current Markdown document only describes the 'high-level' difference count between both the dryrun and the last live run. This is useful for comments on a PR, as the report is concise and outlines the overall effect changes to a given policy, however this is not enough information if you wish to know which resources have been affected by policy changes.

We feel that the inclusion of a detailed HTML report of changes will provide an extra layer of confidence during a review before any policy changes are merged. The mechanism of how to expose this document is entirely up to the user; they could upload it to S3 as a static HTML webpage, integrate it with Slack, email the report, etc.

Below is a brief snapshot of a report - with some information redacted:

![Screenshot 2020-04-28 at 11 03 18](https://user-images.githubusercontent.com/14886082/80475054-35198a80-8940-11ea-98dc-a8dffafd00c6.png)

In order to generate a report, the user will need to have a jinja template `report.j2` located within a folder named `reporting-template` in the same directory where the `dryrun-diff` step has ran.

## Testing Done

Similar to the markdown generation, we log whenever the function is called - however no additional unit testing has been done.

We have modified the mock generation of the BaseStep Class to stub out `self.config.account_id` in order to supplant the mockers issues with `__getattr__`.

## Important Notes

* The [milligram html framework was used for the HTML reports](https://milligram.io/)

## Contributor License Agreement

Required for external contributors.

By submitting this work for inclusion in manheim-c7n-tools, I agree to the following terms:

* The contribution included in this request (and any subsequent revisions or versions of it)
  is being made under the same license as the manheim-c7n-tools project (Apache v2).
* My contribution may perpetually be included in and distributed with manheim-c7n-tools; submitting
  this pull request grants a perpetual, global, unlimited license for it to be used and distributed
  under the terms of manheim-c7n-tools's license.
* I have the legal power and rights to agree to these terms.

JuubiSnake commented on manheim_c7n_tools/dryrun_diff.py:
```diff
@@ -100,6 +109,67 @@ def _find_changed_policies(self, git_dir=None, diff_against='master'):
+        for policy in all_policies:
+            for region in self.config.regions:
+                dry_ids = {
+                    x['Arn']
```
There is a big assumption here regarding the availablity of the `Arn` key... perhaps I'll think of something more clever to extract an ID based on a resource type

jantman replied:
Apologies for the delay reviewing this... I'll get to it next week.

I don't remember offhand, but I'm pretty sure that the c7n data should have a resource ID key in it? If not, there's definitely code in upstream c7n to get that...

JuubiSnake replied:
Good shout, I've opted for the [following imports](https://github.com/manheim/manheim-c7n-tools/pull/34/commits/c9b1b8b77a1c7f16b632d22160fa80d3d12f910c#diff-12aec96e72fb0486abb7d120ef81c3d7R35-R36) from upstream:

```
from c7n.resources import load_resources
from c7n.provider import cloud
```

I'm able to manipuate both packages to generate a ResourceManager for a given `resource_type` which I grab from the prefixes for policies (not sure if that's too big of an assumption). I can then query the ResourceManager to give me the `id` for the associated type; i.e for an `ec2` instance, it'd return `InstanceId`

jantman commented:
@JuubiSnake We just released a new version, 0.10.0, which upgrades to the latest c7n release. Would you mind please rebasing this on master and re-testing? I haven't confirmed, but I _think_ some of the API around ResourceManager and policy loading may have changed...

JuubiSnake commented:
> @JuubiSnake We just released a new version, 0.10.0, which upgrades to the latest c7n release. Would you mind please rebasing this on master and re-testing? I haven't confirmed, but I _think_ some of the API around ResourceManager and policy loading may have changed...

Rebased and added the following https://github.com/manheim/manheim-c7n-tools/pull/34/files#diff-12aec96e72fb0486abb7d120ef81c3d7R35-R36 - turns out that the newer version of c7n makes it a bit eaiser to extract what we need :)

JuubiSnake commented on manheim_c7n_tools/dryrun_diff.py:
```diff
@@ -100,6 +112,87 @@ def _find_changed_policies(self, git_dir=None, diff_against='master'):
+                    policy)
+                continue
+            resource_type = policy_splits[0]
+            load_available()
+            _id = get_resource_class(resource_type) \
+                .resource_type() \
+                .id
```
This _has_ the potential to throw an exception if it cannot find the appropiate resource - perhaps it might be worth placing some guardrails around this...

jantman commented on manheim_c7n_tools/dryrun_diff.py:
```diff
@@ -100,6 +112,87 @@ def _find_changed_policies(self, git_dir=None, diff_against='master'):
+                        %s from report - it is not prefixed correctly",
+                    policy)
+                continue
+            resource_type = policy_splits[0]
```
Assuming that the policy name begins with the resource type isn't going to work for most users, including my own organization.

I know that this will complicate things quite a bit, but alongside the ``resources.json`` files that c7n writes with the policy stats, it also writes the policy itself to a ``metadata.json`` (or ``metadata.json.gz``) file. The right way to find the resource type for each policy would be, in the methods where we get the policy stats (``_get_dryrun_results()`` for the dry-run, and ``_get_s3_results_for_region()`` which calls ``_get_latest_res_for_policy()``) we should also retrieve the metadata files. We can then get the resource type from that.

I know this module is relatively rough code that's not terribly well documented and has no unit tests... so I'll leave it up to you... if you want to try implementing that, cool. If not, I think I'm willing to merge this as-is and then I'll add in that logic myself. Just let me know.

JuubiSnake replied:
Yeah that sounds like a nice solution - I'd be happy to include it in the PR :)

JuubiSnake commented:
> The right way to find the resource type for each policy would be, in the methods where we get the policy stats (_get_dryrun_results() for the dry-run, and _get_s3_results_for_region() which calls _get_latest_res_for_policy()) we should also retrieve the metadata files. We can then get the resource type from that.

@jantman Sorry for the delay - I've included some code for pulling the resourcce_type via `metadata.json` from both the `dryrun` and `live` results. I'll be updating the commit should it fail the tests on travis - but for now, I think it's close to finished. Any feedback you have I'll address 👍

JuubiSnake commented on manheim_c7n_tools/dryrun_diff.py:
```diff
@@ -100,6 +115,100 @@ def _find_changed_policies(self, git_dir=None, diff_against='master'):
+                'getting policy: %s resource type from live run', policy)
+            resource_type = self._live_results.get(policy, {}).get(
+                self.RESOURCE_TYPE_KEY, self.UNKNOWN_RESOURCE_TYPE)
+            if resource_type == self.UNKNOWN_RESOURCE_TYPE:
+                logger.warning(
+                    'unable to get policy: %s resource type from live run', policy)
+                logger.info(
+                    'getting policy: %s resource type from dryrun', policy)
+                resource_type = dryrun.get(policy, {}).get(
+                    self.RESOURCE_TYPE_KEY, self.UNKNOWN_RESOURCE_TYPE)
+                if resource_type == self.UNKNOWN_RESOURCE_TYPE:
+                    logger.error('unable to get policy: %s resource type', policy)
+                    return ''
```
What if we change the resource the policy affects, but not it's name - this might be an issue. I'll work on this.

robertstettner commented:
Hi @jantman, just a friendly bump here. Could you please re-review this, as we are blocked. Thank you.

jantman commented:
@robertstettner apologies, I'd been away from work for the past few days. I'll take another look at this now.

jantman commented:
This looks great, thank you all so much! I'll get it merged now and released ASAP.

jantman commented:
@JuubiSnake @robertstettner This has been released as 1.0.0. Thank you so much! The new version is now live on both PyPI and the Docker Hub.
